torchvision/prototype/datasets/_builtin/food101.py

```diff
@@ -0,0 +1,86 @@
+import functools
+from pathlib import Path
+from typing import Any, Tuple, List, Dict, cast, Optional, BinaryIO
+
+from torchdata.datapipes.iter import IterDataPipe, Filter, Mapper, CSVParser, JsonParser, Demultiplexer
+from torchvision.prototype.datasets.utils import Dataset, DatasetInfo, DatasetConfig, HttpResource, OnlineResource
+from torchvision.prototype.datasets.utils._internal import (
+    hint_shuffling,
+    hint_sharding,
+    path_comparator,
+    getitem,
+    INFINITE_BUFFER_SIZE,
+)
+from torchvision.prototype.features import Label, EncodedImage
+
+
+class Food101(Dataset):
+    def _make_info(self) -> DatasetInfo:
+        return DatasetInfo(
+            "food101",
+            homepage="https://data.vision.ee.ethz.ch/cvl/datasets_extra/food-101",
+            # Only train and test splits, no validation.
+            valid_options=dict(split=("train", "test")),
+        )
+
+    def resources(self, config: DatasetConfig) -> List[OnlineResource]:
+        return [
+            HttpResource(
+                url="http://data.vision.ee.ethz.ch/cvl/food-101.tar.gz",
+                sha256="d97d15e438b7f4498f96086a4f7e2fa42a32f2712e87d3295441b2b6314053a4",
+                decompress=True,
```

Comment:
Did you verify that the compression is actually a bottleneck here? If not could you run a small benchmark like

```py
import time
from torchvision.prototype import datasets, transforms

tic = time.perf_counter()
dataset = datasets.load("food101")
next(iter(dataset))
tac = time.perf_counter()
for sample in dataset.map(transforms.DecodeImage()):
    pass
toc = time.perf_counter()
print("warmup", toc - tic)
print("iter", toc - tac)
```

Note that in order to test the uncompressed variant, you temporarily need to remove the file from the data folder (`~/.cache/torch/datasets/vision/food101`) and remove `decompress=True`.

Reply:
Without the decompress=True, it is taking a very long time (I was expecting few seconds of run time but it is still running even after few minutes) and I am having this warning:


```
/home/yalouini/miniconda3/envs/torchvision/lib/python3.9/site-packages/torch/utils/data/datapipes/iter/combining.py:203: UserWarning: Some child DataPipes are not exhausted when __iter__ is called. We are resetting the buffer and each child DataPipe will read from the start again.
  warnings.warn("Some child DataPipes are not exhausted when __iter__ is called. We are resetting "
```

I will keep the benchmark that you have suggested running and see later how long it took but something looks weird here for sure.

Reply:
The warning is expected, since we start the iteration twice. Here is a tweaked benchmark code:

```py
import time
from torchvision.prototype import datasets, transforms

tic = time.perf_counter()

dataset = iter(datasets.load("food101").map(transforms.DecodeImage()))
next(dataset)

tac = time.perf_counter()

num_samples = 1
for sample in dataset:
    num_samples += 1
    
toc = time.perf_counter()

print(f"warmup: {tac - tic:.2f} seconds")
print(f"iter: {num_samples / (toc - tac):.1f} samples per second")
```

Reply:
I will have some dedicated time this afternoon at work, so will check it, thanks.
I am also resolving comments that have been fixed (first time working on a PR on github, used to gitlab and bitbucket). :ok_hand:

Reply:
No worries :slightly_smiling_face: Thanks for the continued work on this!

Reply:
So, managed to run the warmup check and it is quite long (I guess it should be few seconds at most) =>

`warmup: 98.69 seconds`

What remains (the sample loop) is still running as expected given the warmup time. 

Could it be something related to my computer? I will give it a try on my personal desktop later this weekend and check how it performs. 

Here are some of the specs of the computer running the benchmark:

![computer_specs](https://user-images.githubusercontent.com/8986994/159706996-d17b0d4a-21bf-4dce-87ee-50b5030206d9.png)

Reply:
> So, managed to run the warmup check and it is quite long (I guess it should be few seconds at most) =>

Yes, everything above let's say 5 seconds is worth investigating / fixing.

> warmup: 98.69 seconds

Is this with or without compression? What timing do you get for the other variant?

Reply:
So I did a CIFAR benchmark to have a comparaison point and it seems it is doing fine, so probably not an issue with my computer:

![cifar_benchmark](https://user-images.githubusercontent.com/8986994/159710356-1a9c093d-0e3e-4849-8a0b-b2d693d3996a.png)

I will check how the dataset is coded and report back.

Reply:
Before you go any deeper into this, could you please answer if you are seeing a similar warmup time for compressed and decompressed files? Because this very much looks like a case where decompressing the files is warranted.

Reply:
So, managed to run the warmup check and it is quite long (I guess it should be few seconds at most) =>

Yes, everything above let's say 5 seconds is worth investigating / fixing.

warmup: 98.69 seconds

Is this with or without compression? What timing do you get for the other variant?

That's the time without any flag turned to True. I will set decompress=True and report back.

Reply:
@pmeier Yes, it looks like the dataset is much bigger than the CIFAR10 one (as I was expecting it). The check is in progress, will report as soon as it is done.

Reply:
Alright, it looks much better with the` decompress=True` =>

```
warmup: 12.22 seconds
iter: 673.5 samples per second
```
Will try with `extract=True` enabled as well.

Reply:
With extract=True as well =>

```
warmup: 12.90 seconds
iter: 560.0 samples per second
```
Warmup is about the same and iteration are slightly better. I guess we could push this if we want as much gains as possible.

Reply:
> iteration are slightly better

Nope. More samples per second is better. Thus, `decompress=True` gives the best performance here so let's go with that. The high warm-up time is bad, but probably can't be avoided.

Reply:
Heads-up: after #5669, the flag should be `preprocess="decompress"`.

Reply:
That's great, I will merge the latest main and update the flag. Thanks @pmeier.

Reply:
I'll have a look why we have that much warmup and will get back to you.